File: implementations/javascript/src/hash.mjs
/**
 * VCITE SHA-256 Hash -- Standalone ES Module Reference Implementation
 *
 * Computes passage fingerprints per VCITE spec section 5. The hash algorithm:
 * 1. Normalize each segment: NFC unicode, collapse whitespace, trim
 * 2. Pad context (text_before, text_after) to exactly 50 code points
 * 3. Concatenate: pad(before) + "|" + normalize(exact) + "|" + pad(after)
 * 4. SHA-256 the UTF-8 bytes, output as "sha256:" + lowercase hex
 *
 * Supports both Node.js (crypto module) and browsers (Web Crypto API).
 * Zero external dependencies.
 */

/** Context window size in Unicode code points (spec section 5.1). */
const CONTEXT_LEN = 50;

/**
 * NFC normalize and collapse ASCII whitespace runs to a single space.
 * Per spec: only U+0009 (tab), U+000A (LF), U+000D (CR), U+0020 (space)
 * are collapsed. Non-breaking space (U+00A0) and other Unicode whitespace
 * are intentionally preserved.
 *
 * @param {string} s
 * @returns {string}
 */
export function normalizeSegment(s) {
  let normalized = s.normalize('NFC');
  normalized = normalized.replace(/[\t\n\r ]+/g, ' ');
  return normalized.trim();
}

/**
 * Pad or truncate a context string to exactly `length` Unicode code points.
 * Uses Array.from() for multi-byte character safety (emoji, CJK, etc.).
 *
 * @param {string} s
 * @param {number} [length=50]
 * @returns {string}
 */
export function padContext(s, length = CONTEXT_LEN) {
  const chars = Array.from(s);
  if (chars.length >= length) {
    return chars.slice(0, length).join('');
  }
  return chars.join('') + ' '.repeat(length - chars.length);
}

/**
 * Build the canonical hash input string from the three segments.
 * Format: pad(normalize(before)) + "|" + normalize(exact) + "|" + pad(normalize(after))
 *
 * @param {string} textExact
 * @param {string} [textBefore='']
 * @param {string} [textAfter='']
 * @returns {string}
 */
export function buildHashInput(textExact, textBefore = '', textAfter = '') {
  const before = padContext(normalizeSegment(textBefore));
  const exact = normalizeSegment(textExact);
  const after = padContext(normalizeSegment(textAfter));
  return `${before}|${exact}|${after}`;
}

/**
 * Convert an ArrayBuffer to a lowercase hex string.
 * @param {ArrayBuffer} buffer
 * @returns {string}
 */
function bufferToHex(buffer) {
  const bytes = new Uint8Array(buffer);
  return Array.from(bytes)
    .map(b => b.toString(16).padStart(2, '0'))
    .join('');
}

/**
 * Compute SHA-256 digest of a UTF-8 string.
 * Uses Web Crypto API (Node 18+ and browsers) with Node.js crypto fallback.
 *
 * @param {string} input - UTF-8 string to hash
 * @returns {Promise<string>} lowercase hex digest
 */
async function sha256(input) {
  // Prefer Web Crypto API (available in browsers and Node 18+)
  if (typeof globalThis.crypto !== 'undefined' && globalThis.crypto.subtle) {
    const encoded = new TextEncoder().encode(input);
    const hashBuffer = await globalThis.crypto.subtle.digest('SHA-256', encoded);
    return bufferToHex(hashBuffer);
  }

  // Fall back to Node.js crypto module
  const { createHash } = await import('node:crypto');
  return createHash('sha256').update(input, 'utf8').digest('hex');
}

/**
 * Compute the VCITE passage fingerprint (async, cross-platform).
 *
 * @param {string} textExact - The verbatim cited passage
 * @param {string} [textBefore=''] - Up to 50 chars before the passage
 * @param {string} [textAfter=''] - Up to 50 chars after the passage
 * @returns {Promise<string>} "sha256:" + hex digest
 */
export async function computeHash(textExact, textBefore = '', textAfter = '') {
  const input = buildHashInput(textExact, textBefore, textAfter);
  const digest = await sha256(input);
  return `sha256:${digest}`;
}

File: implementations/python/vcite/hash.py
"""VCITE passage fingerprinting — reference implementation (spec §5.2).

This module implements the SHA-256 passage fingerprint algorithm defined
in the VCITE specification. The algorithm computes a deterministic hash
over a cited passage and its surrounding context, enabling cryptographic
verification that a citation matches its source.

The implementation follows spec §5.1 exactly:
  1. Normalize each segment (NFC + whitespace collapse)
  2. Pad/truncate context windows to exactly CONTEXT_LEN code points
  3. Concatenate with pipe delimiters
  4. SHA-256 over UTF-8 encoding
  5. Prepend "sha256:" prefix

Spec note: The whitespace regex covers U+0009 (tab), U+000A (LF),
U+000D (CR), U+0020 (space) only. Non-breaking space (U+00A0) and
other Unicode whitespace are intentionally NOT collapsed. The context
window counts Unicode code points, not UTF-8 bytes.
"""

import hashlib
import re
import unicodedata

CONTEXT_LEN = 50  # Unicode code points, not bytes


def normalize_segment(s: str) -> str:
    """NFC normalize and collapse ASCII whitespace.

    Per spec §5.1: apply Unicode NFC normalization, then collapse runs
    of tab/LF/CR/space to a single space, then strip leading/trailing
    whitespace from the segment.
    """
    s = unicodedata.normalize("NFC", s)
    s = re.sub(r"[\t\n\r ]+", " ", s)
    return s.strip()


def pad_context(s: str, length: int = CONTEXT_LEN) -> str:
    """Truncate or space-pad to exactly `length` Unicode code points.

    Per spec §5.1: context windows are fixed-length. Shorter strings
    are right-padded with U+0020 spaces. Longer strings are truncated.
    """
    chars = list(s)  # preserves multi-byte chars as single units
    if len(chars) >= length:
        return "".join(chars[:length])
    return "".join(chars) + " " * (length - len(chars))


def compute_hash(
    text_exact: str,
    text_before: str = "",
    text_after: str = "",
) -> str:
    """Compute the VCITE passage fingerprint for a cited passage.

    Args:
        text_exact: The verbatim cited passage.
        text_before: Up to 50 characters of context before the passage.
        text_after: Up to 50 characters of context after the passage.

    Returns:
        A string of the form "sha256:<64 hex chars>".
    """
    before = pad_context(normalize_segment(text_before))
    exact = normalize_segment(text_exact)
    after = pad_context(normalize_segment(text_after))
    raw = f"{before}|{exact}|{after}"
    digest = hashlib.sha256(raw.encode("utf-8")).hexdigest()
    return f"sha256:{digest}"
